lib/sax.js

```diff
@@ -538,86 +538,228 @@ function _copy(source, target) {
+			skipBlanks();
+
+			// Check for External ID and SystemId
+			if (compare('PUBLIC')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between PUBLIC and public identifier');
 				}
-				var lastMatch = matchs[len - 1];
-				domBuilder.startDTD(name, pubid, sysid);
-				domBuilder.endDTD();
+				doctype.publicId = getMatch(grammar.PubidLiteral);
+				if (!doctype.publicId) return errorHandler.fatalError('PUBLIC without public identifier');
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between public identifier and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('PUBLIC without SystemId');
+			} else if (compare('SYSTEM')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between SYSTEM and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('SYSTEM without SystemId');
+			}
+
+			// Parse internal subset
+			skipBlanks();
 
-				return lastMatch.index + lastMatch[0].length;
+			if (char() === '[') {
+				skip(1);
+				var intSubsetStart = index;
+				skipBlanks();
+				while (index < source.length) {
+					var current = null;
+					// Only in external subset
+					// if (char() === '<' && char(1) === '!' && char(2) === '[') {
+					// 	parseConditionalSections();
+					// } else
+					if (char() === '<' && char(1) === '!') {
+						switch (char(2)) {
+							case 'E':
+								if (char(3) === 'L') {
+									current = getMatch(grammar.elementdecl);
+								} else if (char(3) === 'N') {
+									current = getMatch(grammar.EntityDecl);
+								}
+								break;
+							case 'A':
+								current = getMatch(grammar.AttlistDecl);
+								break;
+							case 'N':
+								current = getMatch(grammar.NotationDecl);
+								break;
+							case '-':
+								current = getMatch(grammar.Comment);
+								break;
+						}
+					} else if (char() === '<' && char(1) === '?') {
+						current = parsePI();
+					} else if (char() === '%') {
+						current = getMatch(grammar.PEReference);
+					} else {
+						return errorHandler.fatalError('Error detected in Markup declaration');
+					}
+					if (!current) {
+						return errorHandler.fatalError('Error in internal subset at position ' + index);
+					}
+					skipBlanks();
+					if (char() === ']') {
+						doctype.internalSubset = source.substring(intSubsetStart, index);
+						skip(1);
+						break;
+					}
+				}
+				if (doctype.internalSubset === null) {
+					return errorHandler.fatalError('Error in internal subset: no closing bracket found');
+				}
+			}
+			skipBlanks();
+			if (char() !== '>') {
+				return errorHandler.fatalError('DOCTYPE not terminated');
 			}
+			skip(1);
+			domBuilder.startDTD(doctype.name, doctype.publicId, doctype.systemId, doctype.internalSubset);
+			domBuilder.endDTD();
```

Comment:
In the intended way of sax parsing the methods that would be called for the pieces inside the DTD need to be called after startDTD and before endDTD, which is why the internalSubset can not be passed to the startDTD method when we want to start calling these other methods which are currently no-ops.

But it is a totally valid spproach for this PR.

Reply:
https://github.com/xmldom/xmldom/blob/acf923c9d0853a9fb68e11d4e6ad7c45ad957c59/lib/dom-parser.js#L525-L532

But as I said, not relevant for this PR, so I mark this as resolved